### PR TITLE
CosmWasm smart contract hooks

### PR DESCRIPTION
**Pull request by bigs**
## Description

This closes feature request #50.

First, want to highlight that the primary author of this change is @BurntVal and that my role has merely been to review, give advice, and liaise!

This PR introduces four hooks for interacting with CosmWasm smart contracts in React applications. They are:

- `useInstantiateContract`—a mutator to instantiate a smart contract with a message provided by the caller.
- `useExecuteContract`—a mutator to execute a smart contract, passing it a message provided by the caller.
- `useQueryRaw` and `useQuerySmart`—Hooks to query the state stored within a smart contract.

We made our best effort to follow the style and standards of the graz codebase. We generally made an effort to keep the typing as strong as possible, so the mutators are all parameterized by input and output types.

## Checklist

- [x] I have made sure the upstream branch for this PR is correct
- [x] I have made sure this PR is ready to merge
- [x] I have made sure that I have assigned reviewers related to this project

## Changes

- [x] Added `useInstantiateContract` hook
- [x] Added `useExecuteContract` hook
- [x] Added `useQuerySmart` hook
- [x] Added `useQueryRaw` hook

## Screenshots

n/a

## Testing

To test our changes, we locally extended the sample project in `templates/default` to include queries and transactions made to various sample contracts using all of the hooks added in this change request. Through testing in this application setting, we confirmed instantiation, execution, smart queries, and raw queries all worked.

## Links/References

- [CosmWasm](https://docs.cosmwasm.com/docs/1.0/)
- [@cosmjs/cosmwasm-stargate](https://cosmos.github.io/cosmjs/latest/cosmwasm-stargate/index.html)

## Notes

- In order to make the hooks flexible, many of the hook arguments are optional (in the event that they are not yet known at first component render). In these instances (in the queries, specifically), the hooks will be automatically disabled until they are present to avoid invalid queries.

### CODE REVIEW

**Comment by bigs** [comment]
@codingki This is ready for a review! I'm not sure what's failing in the graz-example deploy, as I've not got permissions to view the project on Vercel.

**Comment by codingki** [comment]
Dont mind the failing builds, we are currently working on it get it fixed @bigs

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
    throw new Error("CosmWasm signing client is not ready");
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
    throw new Error("CosmWasm signing client is not ready");
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
    throw new Error("CosmWasm signing client is not ready");
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: Let's follow `@cosmjs` params order  
```suggestion
export const getQueryRaw = (address: string, keyStr: string): Promise<Uint8Array | null> => {
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: if the address can't be undefined let's make the params on line 190 required

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: if the `address` and `queryMsg` can't be undefined let's make it required 
```suggestion
export const getQuerySmart = async <TData>(address: string, queryMsg: Record<string, unknown>): Promise<TData> => {
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
    throw new Error("CosmWasm signing client is not ready");
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: Let's make `senderAddress` params required

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
  senderAddress: string;
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: let's make `senderAddress` params required

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]:
```suggestion
  senderAddress: string;
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: since `accountAddress` is a dependant on this function let's add it to the key
```suggestion
  const queryKey = ["USE_INSTANTIATE_CONTRACT", onError, onLoading, onSuccess, codeId, accountAddress];
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: Nice work for separating the function🔥

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: to avoid confusion let's change the naming
```suggestion
    onSuccess: (instantiateResult) => Promise.resolve(onSuccess?.(instantiateResult)),
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: since `accountAddress is a dependant on this function let's add it to the key
```suggestion
  const queryKey = ["USE_EXECUTE_CONTRACT", onError, onLoading, onSuccess, contractAddress, accountAddress];
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: to avoid confusion let's change the naming
```suggestion
    onSuccess: (executeResult) => Promise.resolve(onSuccess?.(executeResult)),
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: I don't think we need to type a key

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: I don't think we need to expose `options`.

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]:
```suggestion
  const queryKey = ["USE_QUERY_SMART", address, queryMsg] as const;
  const query = useQuery(queryKey, ({ queryKey: [, _address] }) => getQuerySmart(address, queryMsg),{
      enabled: Boolean(address) && Boolean(queryMsg)
  });

  return query;
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]:
```suggestion
```

Comment (codingki) [reply packages/graz/src/hooks/methods.ts]: We don't need to type a key

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: I don't think we need to expose `options`

Comment (codingki) [inline packages/graz/src/hooks/methods.ts]:
```suggestion
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]:
```suggestion

  const queryKey: QueryRawKey = ["USE_QUERY_RAW", key, address] as const;
  const query = useQuery(queryKey, ({ queryKey: [, _address] }) => getQueryRaw(key, address), {
      enabled: Boolean(address)
  });
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]:
```suggestion
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: I don't think we need to expose `options`
```suggestion
```

**Review by codingki — CHANGES_REQUESTED** [review]
Awesome work @bigs🔥 Thanks for working on this, We are really appreciate your work🫡 
There are some comments that needs to be resolved and suggestions before merging💪

**Comment by codingki** [comment]
> In order to make the hooks flexible, many of the hook arguments are optional (in the event that they are not yet known at first component render). In these instances (in the queries, specifically), the hooks will be automatically disabled until they are present to avoid invalid queries.

We can make the hooks params optional and have a `enabled` options on the query options

File: packages/graz/src/actions/methods.ts
Comment (bigs) [reply packages/graz/src/actions/methods.ts]: Definitely open to this, but wanted to elaborate a bit more on the reasoning just to be be totally clear (and sorry if this is stuff you're aware of!)—

React requires that hooks be instantiated consistently between renders—the same set of hooks must be called in the same order every time the page renders, or React will give a warning and potentially exhibit undefined behavior. This means that if an argument is required, it must be available/defined within the component using the hook from the very first render.

Having optional arguments like this allows us to use the hook in components where the required data (aka contract address or query message) is made available dynamically, rather than having to nest a sub-component that is conditionally rendered when that data is made available. In my experience, it reduces some cruft. Further, it doesn't preclude the user from taking the sub-component approach, either.

The obvious advantage of having everything required is it makes the logic quite a bit simpler and the implementation more straightforward. I'm truly happy to proceed like that, just wanted to share the reasoning first!

Comment (bigs) [reply packages/graz/src/actions/methods.ts]: The primary use case within `react-query` is [dependent queries](https://tanstack.com/query/v4/docs/guides/dependent-queries)

Comment (codingki) [reply packages/graz/src/actions/methods.ts]: Yes we can still keep that behavior in hooks level, by using `enabled` configuration. in action level we can make them required but in hooks level it's optional

example:
```ts
const getSomething = (foo: string, bar: string) => {
	// foo and bar required in fetchSomething
	return await fetchSomething(foo, bar)
}

const useSomething = ({foo,bar}: {foo? : string, bar? : string}) => {
	const query = useQuery(["USE_SOMETHING", foo, bar], async () => {
                if(!foo || !bar) throw new Error("foo or bar undefined")
		return await getSomething(foo, bar)
	}, {
		enabled: Boolean(foo) && Boolean(bar)
	})
	return query
}
```

hooks still can be called when `foo` & `bar` undefined, but will execute only if `foo` & `bar` is there and `foo` & `bar` changed

Comment (bigs) [reply packages/graz/src/actions/methods.ts]: Seems like a more than reasonable approach :)

**Comment by BurntVal** [comment]
@codingki The requested changes were implemented and is ready for review again :)

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: we can remove this block
```suggestion
```

File: packages/graz/src/actions/methods.ts
Comment (codingki) [inline packages/graz/src/actions/methods.ts]: We can remove this block
```suggestion
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: The code explain it self, remove this comment
```suggestion
```

File: packages/graz/src/hooks/methods.ts
Comment (codingki) [inline packages/graz/src/hooks/methods.ts]: The code explain it self, remove this comment
```suggestion
```

**Review by codingki — APPROVED** [review]
Looks good to me! found couples suggestion non blocking the merge. Great job🫡

**Comment by BurntVal** [comment]
Sorry for any inconvenience. It's the little things that always escape my mind. That should be everything now :)

**Comment by codingki** [comment]
Awesome🔥

**Comment by codingki** [comment]
These will be included in 0.0.28 and up

**Comment by BurntVal** [comment]
Awesome. Thank you guys for the support and quick responses!

**Pull request merged**
